File: host/src/contract.ts
// µhost — contract.ts: versioned tool contracts decoupled from implementations (kernel
// requirement #4) with per-execution generation pinning (#9). Host-side per D-340:
// resolve() sits on the same hot dispatch path as graph.ts::whoOffers — an in-flight
// caller must never be yanked to a newer, possibly-incompatible generation mid-call.
// A dispatch frame holds its resolved target for the full await: that held reference
// IS the Ω GenerationPin. Generations are APPEND-ONLY — publishing never removes a
// prior one, so retirement is a separate explicit decision, never a side effect.
// Range grammar v1: Ω's existing DependencyRef convention ("1.x" prefix | "*" | exact
// version) — no second range grammar is introduced (D-340 §1.4).
export interface GenerationPin { version: string; impl: string }

export class ToolRegistry {
  private generations = new Map<string, GenerationPin[]>();

  publish(name: string, version: string, impl: string): void {
    const list = this.generations.get(name) ?? [];
    list.push({ version, impl });
    this.generations.set(name, list);
  }

  /** Newest generation matching the caller's declared range. */
  resolve(name: string, range: string): GenerationPin | undefined {
    const list = this.generations.get(name);
    if (!list) return undefined;
    for (let i = list.length - 1; i >= 0; i--) if (matchesRange(list[i].version, range)) return list[i];
    return undefined;
  }

  latest(name: string): GenerationPin | undefined { const l = this.generations.get(name); return l?.[l.length - 1]; }
  generationCount(name: string): number { return this.generations.get(name)?.length ?? 0; }
}

export function matchesRange(version: string, range: string): boolean {
  if (range === "*" || range === "" || range === "latest") return true;
  if (range.endsWith(".x")) return version.split(".")[0] === range.slice(0, -2); // "1.x" matches any 1-series ("1", "1.2")
  return version === range;
}

File: docs/parity/vivim_omega_core-reference/src/contract.rs
//! Versioned tool contracts, decoupled from implementation —
//! requirement #4 — plus per-execution generation pinning —
//! requirement #9.
//!
//! A tool exposes a stable, named interface at a specific semantic
//! version. Plugins never depend on a specific implementation build —
//! they declare a `VersionReq` range against the interface name (see
//! manifest.rs). This is what lets a tool's internals be rewritten
//! completely as long as the contract still holds, and it's what lets
//! old and new callers coexist across an upgrade instead of forcing a
//! flag-day migration.

use semver::{Version, VersionReq};
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Debug, Clone)]
pub struct Contract {
    pub name: String,
    pub version: Version,
}

/// One generation of a tool's implementation. Wrapped in `Arc` so a
/// generation stays alive for exactly as long as something in flight
/// is still using it — no global mutable "current version" pointer,
/// so upgrading never requires stopping the world.
pub struct GenerationPin<T> {
    pub contract: Contract,
    pub implementation: Arc<T>,
}

pub struct ToolRegistry<T> {
    /// Every generation ever published for a tool name, oldest first.
    generations: HashMap<String, Vec<GenerationPin<T>>>,
}

impl<T> Default for ToolRegistry<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> ToolRegistry<T> {
    pub fn new() -> Self {
        Self {
            generations: HashMap::new(),
        }
    }

    /// Publish a new generation. Existing generations are never
    /// removed by this call — retiring an old generation is a
    /// separate, explicit decision (e.g. once its blast radius, per
    /// centrality.rs, drops to zero).
    pub fn publish(&mut self, name: &str, version: Version, implementation: T) {
        self.generations
            .entry(name.to_string())
            .or_default()
            .push(GenerationPin {
                contract: Contract {
                    name: name.to_string(),
                    version,
                },
                implementation: Arc::new(implementation),
            });
    }

    /// Resolve a caller's version requirement against the newest
    /// generation that still satisfies it. A caller that started work
    /// against an older compatible range gets a live `Arc` to exactly
    /// that generation, even after a newer one has been published —
    /// this is the mechanism behind requirement #9.
    pub fn resolve(&self, name: &str, req: &VersionReq) -> Option<Arc<T>> {
        self.generations
            .get(name)?
            .iter()
            .rev()
            .find(|g| req.matches(&g.contract.version))
            .map(|g| g.implementation.clone())
    }

    pub fn latest(&self, name: &str) -> Option<Arc<T>> {
        self.generations.get(name)?.last().map(|g| g.implementation.clone())
    }

    /// How many published generations of a tool are still registered.
    /// A widening gap between `latest` and the oldest generation still
    /// being `resolve`d by live callers is itself a signal worth
    /// watching — it means an upgrade hasn't fully propagated.
    pub fn generation_count(&self, name: &str) -> usize {
        self.generations.get(name).map(|v| v.len()).unwrap_or(0)
    }
}
